plugins/parsers/csv/parser.go

```diff
@@ -17,33 +19,100 @@ import (
 type TimeFunc func() time.Time
 
 type Config struct {
-	ColumnNames       []string `toml:"csv_column_names"`
-	ColumnTypes       []string `toml:"csv_column_types"`
-	Comment           string   `toml:"csv_comment"`
-	Delimiter         string   `toml:"csv_delimiter"`
-	HeaderRowCount    int      `toml:"csv_header_row_count"`
-	MeasurementColumn string   `toml:"csv_measurement_column"`
-	MetricName        string   `toml:"metric_name"`
-	SkipColumns       int      `toml:"csv_skip_columns"`
-	SkipRows          int      `toml:"csv_skip_rows"`
-	TagColumns        []string `toml:"csv_tag_columns"`
-	TimestampColumn   string   `toml:"csv_timestamp_column"`
-	TimestampFormat   string   `toml:"csv_timestamp_format"`
-	Timezone          string   `toml:"csv_timezone"`
-	TrimSpace         bool     `toml:"csv_trim_space"`
-	SkipValues        []string `toml:"csv_skip_values"`
+	ColumnNames        []string `toml:"csv_column_names"`
+	ColumnTypes        []string `toml:"csv_column_types"`
+	Comment            string   `toml:"csv_comment"`
+	Delimiter          string   `toml:"csv_delimiter"`
+	HeaderRowCount     int      `toml:"csv_header_row_count"`
+	MeasurementColumn  string   `toml:"csv_measurement_column"`
+	MetricName         string   `toml:"metric_name"`
+	SkipColumns        int      `toml:"csv_skip_columns"`
+	SkipRows           int      `toml:"csv_skip_rows"`
+	TagColumns         []string `toml:"csv_tag_columns"`
+	TimestampColumn    string   `toml:"csv_timestamp_column"`
+	TimestampFormat    string   `toml:"csv_timestamp_format"`
+	Timezone           string   `toml:"csv_timezone"`
+	TrimSpace          bool     `toml:"csv_trim_space"`
+	SkipValues         []string `toml:"csv_skip_values"`
+	MetadataRows       int      `toml:"csv_metadata_rows"`
+	MetadataSeparators []string `toml:"cvs_metadata_separators"`
+	MetadataTrimSet    string   `toml:"cvs_metadata_trim_set"`
+
+	MetadataSeparatorList MetadataPattern
 
 	gotColumnNames bool
 
-	TimeFunc    func() time.Time
-	DefaultTags map[string]string
+	TimeFunc     func() time.Time
+	DefaultTags  map[string]string
+	MetadataTags map[string]string
 }
 
 // Parser is a CSV parser, you should use NewParser to create a new instance.
 type Parser struct {
 	*Config
 }
 
+type MetadataPattern []string
+
+func (record MetadataPattern) Len() int {
+	return len(record)
+}
+func (record MetadataPattern) Swap(i, j int) {
+	record[i], record[j] = record[j], record[i]
+}
+func (record MetadataPattern) Less(i, j int) bool {
+	//if len(record[i]) == len(record[j]) {
+	//	return record[i] < record[j]
+	//}
+	return len(record[i]) > len(record[j])
+}
```

Comment:
Hmm `Less` should return `true` if `i` is "less" than `j`. In this context it should return `true` if the record in `i` is _shorter_ than the one in `j`. I guess you are doing it this way as you need the _reverse_ sorting by length right? Anyway, a comment would help for others to understand this.

Furthermore, please delete the commented, dead code.